tools/src/main/java/org/apache/kafka/tools/consumer/CoordinatorRecordMessageFormatter.java

```diff
@@ -31,44 +34,50 @@
 
 import static java.nio.charset.StandardCharsets.UTF_8;
 
-public abstract class ApiMessageFormatter implements MessageFormatter {
-
+public abstract class CoordinatorRecordMessageFormatter implements MessageFormatter {
     private static final String TYPE = "type";
     private static final String VERSION = "version";
     private static final String DATA = "data";
     private static final String KEY = "key";
     private static final String VALUE = "value";
-    static final String UNKNOWN = "unknown";
+
+    private final CoordinatorRecordSerde serde;
+
+    public CoordinatorRecordMessageFormatter(CoordinatorRecordSerde serde) {
+        this.serde = serde;
+    }
 
     @Override
     public void writeTo(ConsumerRecord<byte[], byte[]> consumerRecord, PrintStream output) {
+        if (Objects.isNull(consumerRecord.key())) return;
+
         ObjectNode json = new ObjectNode(JsonNodeFactory.instance);
+        try {
+            CoordinatorRecord record = serde.deserialize(
+                ByteBuffer.wrap(consumerRecord.key()),
+                consumerRecord.value() != null ? ByteBuffer.wrap(consumerRecord.value()) : null
+            );
+
+            if (!isRecordTypeAllowed(record.key().apiKey())) return;
 
-        byte[] key = consumerRecord.key();
-        if (Objects.nonNull(key)) {
-            short keyVersion = ByteBuffer.wrap(key).getShort();
-            JsonNode dataNode = readToKeyJson(ByteBuffer.wrap(key));
+            json
+                .putObject(KEY)
+                .put(TYPE, record.key().apiKey())
```

Comment:
`ShareGroupStateMessageFormatter` use "version" rather than "type", but I guess we fix it in the follow-up by refactoring `ShareGroupStateMessageFormatter`?

Reply:
That's right. We will it in https://github.com/apache/kafka/pull/18510.